Add error message to quarkus-resteasy-reactive-links and improve documentation

leo-bogastry opened the pull request:
This PR was triggered by the issue #36463

Goal is to clarify the error message the user gets when the class returned by the endpoint does not have and `id` field and also the documentation about [web links](https://quarkus.io/guides/resteasy-reactive#web-links-support)

gsmet commented on extensions/resteasy-reactive/quarkus-resteasy-reactive-links/runtime/src/main/java/io/quarkus/resteasy/reactive/links/runtime/GetterAccessorsContainer.java:
```diff
@@ -11,7 +11,12 @@ public class GetterAccessorsContainer {
+            return getterAccessors.get(className).get(fieldName);
+        } catch (NullPointerException ex) {
+            throw new IllegalStateException("Cannot generate web links for " + className
+                    + " because the class does not contain an `id` field");
```
Shouldn't it be:
```suggestion
                    + " because the class does not contain a field named `" + fieldName + "`");
```
?
It looks more future-proof.

LeonorLunatech replied:
But the issue if not `fieldName` missing, is that is it always expected a field `id` to be present. 
I don't know who came up with the rules, but that's the current implementation.

gsmet replied:
OK, so I'm the one who is confused :).

I think what confuses me is that catching the exception here is a bit of an anti-pattern. Is there any way to catch this condition earlier when inspecting the class instead of that late when getting an unrelated field name?

Note that I haven't looked at the code at all so I'm giving this advice just by inspecting the diff. I might be completely wrong :).

LeonorLunatech replied:
I tried to catch this in the deployment module, in `LinksProcessor.implementPathParameterValueGetters`, after setting the data in `getterAccessorsContainer`.
I tried it by adding a `contains` method to the `GetterAccessorsContainer` and checking if it contains the fieldname `id`, but I got an error  because it involves calling a `getValue`: `You cannot invoke getValue() directly on an object returned from the bytecode recorder, you can only pass it back into the recorder as a parameter`.
I'm totally open for a better implementation. I'm also not very fond of this one

gsmet replied:
Hey, sorry for the late reply, I was on PTO.

Yes, so this all boils down to Quarkus architecture. The build steps are executed at build time and you can record (via a recorder) things that will be executed at runtime (either at static init or runtime init). A value that is initialized by a recorder (the `RuntimeValue` you have here) is a pointer to what you will obtain later at runtime. You can't do anything with it in the build steps... but you can pass it to a recorder method.

So I would recommend you add a new method to the recorder to validate things and call it at the end of `implementPathParameterValueGetters`. This is very similar to what is already done in this method (typically with `getterAccessorsContainerRecorder.addAccessor` for instance).

Let me know if something is unclear, I will be happy to help.

gsmet reviewed (commented):
Thanks. I added a very small comment about the error message.

gsmet commented on extensions/resteasy-reactive/quarkus-resteasy-reactive-links/deployment/src/main/java/io/quarkus/resteasy/reactive/links/deployment/LinksProcessor.java:
```diff
@@ -188,6 +188,8 @@ private RuntimeValue<GetterAccessorsContainer> implementPathParameterValueGetter
                     }
                 }
+
+                getterAccessorsContainerRecorder.validateContainer(getterAccessorsContainer, entityType);
```
Wouldn't it be easier to move this one line below and validate all the entities at once? Also you could list all the entities without ids in one go in the exception you throw? Otherwise it's going to be a bit tedious when you have several entities without the id.

All in all, I think it should also simplify your code (unless I'm missing something).

leo-bogastry replied:
I was keeping track of the `entityType`
I pushed a modification of the validation to check all the container only once and fixed the test

gsmet commented on extensions/resteasy-reactive/quarkus-resteasy-reactive-links/deployment/src/test/java/io/quarkus/resteasy/reactive/links/deployment/RestLinksWithFailureInjectionTest.java:
```diff
@@ -0,0 +1,31 @@
+    @RegisterExtension
+    static final QuarkusUnitTest TEST = new QuarkusUnitTest()
+            .withApplicationRoot((jar) -> jar
+                    .addClasses(TestRecordNoId.class, TestResourceNoId.class));
```
I think what you're looking for in this case is `QuarkusUnitTest#assertException()`.

And in your `@Test` method, just do a `fail()` as it should never go there.

leo-bogastry replied:
Great! That worked

gsmet reviewed (commented):
I added some additional guidance. HTH.

FroMage commented on extensions/resteasy-reactive/quarkus-resteasy-reactive-links/deployment/src/main/java/io/quarkus/resteasy/reactive/links/deployment/LinksProcessor.java:
```diff
@@ -188,7 +188,9 @@ private RuntimeValue<GetterAccessorsContainer> implementPathParameterValueGetter
                 }
+
             }
+            getterAccessorsContainerRecorder.validateContainer(getterAccessorsContainer);
```
I don't understand why we're delaying the validation check until startup via a recorder. We can validate this in the build phase, no?

leo-bogastry replied:
I though this was the build phase. If not, then I honestly don't know where that is 😅 
Either way the validation I added is specifically for a field called `id`.  But given that just a field annotated with `@Id` will also work, I'm not seeing how that can be validated :/

FroMage commented on docs/src/main/asciidoc/resteasy-reactive.adoc:
```diff
@@ -1619,7 +1619,34 @@ To enable Web Links support, add the `quarkus-resteasy-reactive-links` extension
+----
+public class Record {
+
+    // id is a mandatory field for the generation of web links
```
Looking at the implementation, this does not appear to be true. ATM it's limited in that the endpoint `@PathParam` parameter must be named `id`, but the field from the entity has to be annotated with `@Id`, if I understand it correctly. This doesn't appear to be documented, though, so further testing should be done to validate this.

leo-bogastry replied:
The original issue comes with a project to reproduce the problem and there the endpoint that throws the exception is defined with an `@PathParam("id")` already.
I tested adding the @jakarta.persistence.Id to the field in the `TestRecord` class (called `name`) and the endpoint works properly that way.
So what seem to be needed for the web links to work is either a field called `id` or an @Id annotation (that ends with `persistence.Id`)

FroMage reviewed (changes requested):
Added some questions after looking at the code.

leo-bogastry commented:
@gsmet @FroMage I'm taking another look at this and I am able to get all the fields and annotations of a class (and then search for the `id` or `@Id`) with this code snippet:
```
private void validateClassHasFieldId(IndexView index, DotName className) {
        List<FieldInfo> allFields = index.getClassByName(className).fields();
        List<AnnotationInstance> allAnnotations = index.getClassByName(className).annotations();
...
```

but this does not include the fields and annotations a class may inherit from another one.
Is there a way of accessing that? (I was only able to find a `getAllKnownSubclasses` method an similar)

leo-bogastry commented:
@gsmet @FroMage I took another look at this today and I found the information I was missing.
I updated the code to perform a validation *before* the creation of the web links that searches for fields named `id` or fields annotated with `@Id` in the current class and super classes.
In the tests I updated the `TestRecord` with another level of inheritance and also updated the documentation

FroMage commented on extensions/resteasy-reactive/quarkus-resteasy-reactive-links/deployment/src/main/java/io/quarkus/resteasy/reactive/links/deployment/LinksProcessor.java:
```diff
@@ -194,6 +198,43 @@ private RuntimeValue<GetterAccessorsContainer> implementPathParameterValueGetter
+    private void validateClassHasFieldId(IndexView index, String entityType) {
+        // create a new independent class name that we can override
+        DotName className = DotName.createSimple(entityType);
+        List<FieldInfo> classFields = index.getClassByName(className).fields();
```
You should cache the result of `index.getClassByName()` instead of calling it multiple times here and below.

FroMage replied:
I think a better strategy would be to ask for fields named `id` and for annotations of the `jakarta.persistence.Id` type, rather than building a list of all fields and annotations. You can do this for each class in the hierarchy until you find one or you reach `Object`.

FroMage commented on extensions/resteasy-reactive/quarkus-resteasy-reactive-links/deployment/src/main/java/io/quarkus/resteasy/reactive/links/deployment/LinksProcessor.java:
```diff
@@ -194,6 +198,43 @@ private RuntimeValue<GetterAccessorsContainer> implementPathParameterValueGetter
+        // also get fields and annotations from all super classes
+        DotName superClassName = index.getClassByName(className).superName();
+        while (superClassName != null) {
+            List<FieldInfo> allSuperFields = index.getClassByName(superClassName).fields();
```
This `index.getClassByName(superClassName)` can return `null` if the superclass is not indexed, you should guard for that.

leo-bogastry replied:
@FroMage I added some more guards for the null cases and rewrote the code using recurrency, I think it really improves readability in this case

FroMage reviewed (changes requested):
This is much better already, because it moved the validation to build time. There are a few details remaining to fix, but we're almost there :)

FroMage commented on extensions/resteasy-reactive/quarkus-resteasy-reactive-links/deployment/src/main/java/io/quarkus/resteasy/reactive/links/deployment/LinksProcessor.java:
```diff
@@ -194,6 +197,52 @@ private RuntimeValue<GetterAccessorsContainer> implementPathParameterValueGetter
+                .filter(f -> f.name().equals("id"))
+                .collect(Collectors.toList());
+        List<AnnotationInstance> fieldsAnnotatedWithId = classInfo.annotations().stream()
+                .filter(a -> a.name().toString().endsWith("persistence.Id"))
```
This is looking for annotations found in the class, which includes field annotations, but also method annotations, if you are looking for field annotations you should filter them. Note that you can also find them by traversing the fields, otherwise.

FroMage commented on extensions/resteasy-reactive/quarkus-resteasy-reactive-links/deployment/src/main/java/io/quarkus/resteasy/reactive/links/deployment/LinksProcessor.java:
```diff
@@ -194,6 +197,52 @@ private RuntimeValue<GetterAccessorsContainer> implementPathParameterValueGetter
+
+        // Id field not found and hope is gone
+        DotName superClassName = classInfo.superName();
+        if (fieldsNamedId.isEmpty() && fieldsAnnotatedWithId.isEmpty() && superClassName == null) {
```
```suggestion
        if (superClassName == null) {
```
The first two tests are necessarily always true, because they're already tested above with the `return` case.

FroMage commented on extensions/resteasy-reactive/quarkus-resteasy-reactive-links/deployment/src/main/java/io/quarkus/resteasy/reactive/links/deployment/LinksProcessor.java:
```diff
@@ -194,6 +197,52 @@ private RuntimeValue<GetterAccessorsContainer> implementPathParameterValueGetter
+
+        // Id field not found but there's still hope
+        classInfo = index.getClassByName(superClassName);
+        if (superClassName != null && classInfo != null) {
```
So, what happens if we have a `superClassName` but the superclass is not indexed? No error?

leo-bogastry replied:
How can I test for that scenario?

leo-bogastry replied:
Indeed, nothing would happen in that scenario. I added a check and exception for this case.

FroMage reviewed (changes requested):
Sorry for the late review, this is getting better but there are still things to tweak.

FroMage reviewed (approved):
Well, all comments were resolved, so this is looking good :)

FroMage commented:
Review is OK, but can you rebase and remove merge commits please? Then CI will tell us.

leo-bogastry committed:
Add error message to quarkus-resteasy-reactive-links when there's missing configuration and improve documentation

leo-bogastry commented:
> Review is OK, but can you rebase and remove merge commits please? Then CI will tell us.

@FroMage PR all cleaned up 🧹

FroMage commented:
Alright, done, thanks a lot for your patience !
